src/core/namespace.c

```diff
@@ -869,6 +935,28 @@ static int mount_tmpfs(const MountEntry *m) {
         return 1;
 }
 
+static int mount_mount_path(const MountEntry *m, const char *root_directory) {
+        unsigned long flags = m->flags;
+        assert(m);
+
+        if (!root_directory)
+                root_directory = mount_entry_path(m);
+
+        /* Unlike other mounts, use MS_RDONLY already during first mount pass, in case the source is read-only */
+        if (mount_entry_read_only(m))
+                flags |= MS_RDONLY;
+
+        /* First, get rid of everything that is below if there is anything. Then, overmount with our image */
+
+        (void) mkdir_p_label(root_directory, 0755);
+        (void) umount_recursive(root_directory, 0);
+
+        if (mount(mount_entry_source(m), root_directory, m->fstype_malloc, flags, mount_entry_options(m)) < 0)
+                return log_debug_errno(errno, "Failed to mount %s: %m", root_directory);
```

Comment:
hmm, so the problem with this is that various file systems require userspace callouts for this. i.e. exfat, ntfs, zfs all do. This means unless you know the used file system very well just doing mount() is not sufficient, we'd actually have to call /bin/mount. 

I think we really should use the dissect logic here however, which sidesteps the issue a bit, because we can declare what images have to do to be compatible with this, and that includes only supporting a subset of file systems (we currently don't document the subset of supported file systems, but really should).

Reply:
Another issue (and the reason why I have not been developing this) is that there's also the fundamental kernel mounting logic that one device can't be mounted in writable mode more than once. I forgot this when I was planning to use MountPaths for mounting `/home` for both `display-manager.service` and `user@.service`, which was one of my original goals. So I'm not sure if allowing random mount operations is the right approach after all. It would be a powerful tool though.

Reply:
Hi, shameless plug: I'd _really_ like to use this feature with dm-verity volumes, which would be much simpler with /bin/mount or libmount since I added verity options there recently. mount() doesn't understand it so pre-processing would be required. See https://github.com/systemd/systemd/issues/14697 for details on the use case. Of course I'm happy to iterate on this in the future myself if there's no time/interest right now. Just though it could be interesting to hear a different use case.
Thanks for working on this feature!

Reply:
@bluca note that RootImage='s backend code actually supports dm-verity perfectly. We do not expose an interface to configure or discvoer the root hash however. Would be happy to add that!

Reply:
I suppose libmount should be used in place of raw mount.

I'm also interested to use dm-verity for my squashfs RootImage volumes, but I have not been successful.

Reply:
i am not sure if libmount/forking off external helpers would be that ideal for RootImage= and friends. i.e. i am not sure it's worth supporting zfs and so on for this...

Reply:
@poettering for RootImage (and PathImage too) and verity here's a few things I'd like to contribute support for:

1) standalone hashdevice
2) roothash signature
3) reusing device mapper if already open with same source/hash/signature

If your preference is for reimplementing from scratch in dissect-image, I can do that - it basically means copy&paste from https://github.com/karelzak/util-linux/blob/master/libmount/src/context_veritydev.c - on the other hand, using libmount and the mnt_context_* APIs would result in a lot of code removal but also require a bit of refactoring, as right now things like creating nodes and opening devices is done manually and kind-of separately from the final mounting.

I'll start with reimplementing, and if you decide on refactoring later it can be done.

In terms of configuring, I'm actually fine with the current way of having foo.raw (but would like to have foo.squashfs too) and foo.roothash (and would then add foo.hashdevice and foo.p7). If there were explicit options, how would it look like? Appending somehow to RootImage, or new parameters - eg: RootImageRootHash RootImageHashDevice RootImageHashSignature?

Reply:
bug fix to allow RootImage with squashfs + verity https://github.com/systemd/systemd/pull/16046

Reply:
@bluca note that the dissect logic can already set up verity for you, it however requires that the image file contains the verity data too, i.e. that you have a GPT envelope that contains two partitions at least: the root fs and the verity data for it. If you want to support detached verity data and naked file systems (i.e. lacking a GPT envelope) then I figure we should add two things: options to explicitly configure root hash and path to verity data file. i.e. introduce RootHash= and RootVerity= or so alongside the existing RootImage=. And then maybe auto discovery: if RootImage= is used look for a file <image>.roothash in the same for the hash and <image>.verity for the verity data. In fact nspawn already does the root hash discovery like that. But I figure this should be generalized and moved out of npsawn and into the dissector. 

Either way, if you want to configure these things separately then dissect-image.c needs improvements to take these parameters separately.

Reply:
In the linked PR I added autodiscovery of .hashdevice, and command line parameters to override it to nspawn and systemd-dissect, and support for external hash device when calling cryptsetup.

I will add RootVerity and RootHash (and RootSignature) in the next PR as the next step.

Reply:
Here's a PR to add RootHash and RootVerity service options: https://github.com/systemd/systemd/pull/16132

Reply:
I published the Makefile script I've used for building RootImages as [topimiettinen/make-images](https://github.com/topimiettinen/make-images).

Reply:
@topimiettinen I have pushed a branch on my fork with your commit, plus another to apply the changes suggested by @poettering and switch to use the dissect-image logic instead of manual mount() (and also renamed from PathImages to MountImages again as suggested).
When you have time, could you please have a quick look and see if the end result if of your liking? dm-verity is now supported on single-filesytem images too, which IIRC is one of the features you were after. Thanks!

https://github.com/bluca/systemd/commit/1557276d7bd0a7ab

Reply:
@bluca Thanks for finishing the work and I'm looking forward to try dm-verity support.

The patches look OK to me. Now that raw mount options are not allowed anymore, MountImages probably won't be useful for me, but I hope it will be useful for you and others.

Reply:
@topimiettinen what raw mount options were you intending to use and how? Given dissect-image seems the way to go to handle volumes, I'm sure we can work out a way to support more use cases. For example the third parameter could be added back for this option, and it could be wired into the dissect-image logic.

Reply:
@bluca At least suid/nosuid (so RootImage can start implying nosuid), exec/noexec, dev/nodev, sync. To handle special cases, file system specific flags (devpts: newinstance, ext4: discard, etc). Also it should be possible to use MountImage as filesystem root (so it can be used as RootImage but with arbitrary mount flags), I think you removed handling of that special case. The idea was to construct the complete file system for a service (especially for user applications) from scratch with individual mounts with finely tuned mount flags, without any restrictions or interference from systemd. But now I think my use case (sandbox userland applications) can be better served by mounting the file systems somewhere with mount units and then using BindPaths= to use the whole tree. This handles the problem with read-write mounts better as well.

Reply:
@topimiettinen I see - the reason I removed the option to use MountImage as RootImage was the review comment from Lennart about duplication. What if both RootImage and MountImages supported flags as the last parameter in the colon separated string? Extending the existing directive seems like a backward-compatible change, so in principle it should be doable.

Interestingly, right now in my use case I use mount units to construct the application filesystem view, and then use BindPaths - it works, but not very well. Very finicky, lots of visible moving parts, and https://github.com/systemd/systemd/issues/10872 is an absolute pain to deal with (I actually had to change all mount units to oneshot services, it was just unusable otherwise). That's one of the reasons I'm working hard to be able to move to RootImage + this new directive exclusively.

Reply:
Extra parameter would be fine for me too.

But this is interesting indeed :smiley: I thought your approach is better and vice versa. Perhaps neither is the right one then?

How about some kind of declarative approach? Something like systemd-repart.d but instead of formatting instructions, it would contain instructions on how to mount something somewhere?

Reply:
Very much possible, but also the environment/usecase is likely different, so different considerations might both be correct - I do have a few months of being unhappy with the model in production-testing, though :-D

A declarative approach would be quite interesting indeed - these options could be the building blocks used under the hood. I'll give this some thinking as well while I work on these PRs.

Reply:
Opened https://github.com/systemd/systemd/pull/16321 with the reworked feature, without the mount options. Opened https://github.com/systemd/systemd/pull/16308 to add mount options to RootImage, and once that is sorted, will follow up with the same mechanism as it is agreed for MountImages